src/undocommand.js

```diff
@@ -0,0 +1,192 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Command from '../command/command.js';
+
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * undo.UndoCommand
+ */
+export default class UndoCommand extends Command {
+	/**
+	 * @see engine.command.Command
+	 * @param {engine.Editor} editor
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Batches which are saved by the command. They can be reversed.
+		 *
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
+
+		/**
+		 * Stores the selection ranges for each batch and use them to recreate selection after undo.
+		 *
+		 * @private
+		 * @member {WeakMap.<engine.treeModel.Range>} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
```

Comment:
Also keep in mind, that `Selection` uses `LiveRange`s so it would update whenever document is updated which is not what we want.